### PR TITLE
PEP489 support: multi-phase module initialization, heap based types

### PR DESCRIPTION
**Pull request by tseaver**
Continuation of PR #301, after moving the branch to this repository.

### CODE REVIEW

**Review by dataflake — COMMENTED** [review]
Please don't include me in reviews for C code, I just can't help there.

**Comment by tseaver** [comment]
@dataflake Sorry about that!

File: src/zope/interface/_zope_interface_coptimizations.c [hunk 1]
```diff
@@ -365,315 +263,447 @@ Spec_clear(Spec* self)
...
+    .name               = OSD__name__,
+    .basicsize          = 0,
+    .flags              = LEAFTYPE_FLAGS,
+    .slots              = OSD_type_slots
+};
+
+#endif
+
+/*
+ *  ClassProviderBase class
```
Comment (davisagli) [inline src/zope/interface/_zope_interface_coptimizations.c:537]:
```suggestion
 *  ClassProvidesBase class
```

File: src/zope/interface/_zope_interface_coptimizations.c [hunk 2]
```diff
@@ -1019,133 +1002,122 @@ IB_richcompare(IB* self, PyObject* other, int op)
...
+    {Py_tp_init,        IB__init__},
+    {Py_tp_hash,        IB__hash__},
+    {Py_tp_richcompare, IB_richcompare},
+    {Py_tp_call,        IB__call__},
+    {Py_tp_traverse,    IB_traverse},
+    {Py_tp_clear,       IB_clear},
+    {Py_tp_dealloc,     IB_dealloc},
+    {Py_tp_methods,     IB_methods},
+    {Py_tp_members,     IB_members},
+    /* tp_base cannot be set as a stot -- pass to PyType_FromModuleAndSpec */
```
Comment (davisagli) [inline src/zope/interface/_zope_interface_coptimizations.c:1066]:
```suggestion
    /* tp_base cannot be set as a slot -- pass to PyType_FromModuleAndSpec */
```

File: src/zope/interface/_zope_interface_coptimizations.c [hunk 3]
```diff
@@ -1019,133 +1002,122 @@ IB_richcompare(IB* self, PyObject* other, int op)
...
+    PyObject* _cache;
+    PyObject* _mcache;
+    PyObject* _scache;
+} LB;
 
-  return 0;
+static int
+LB_traverse(LB* self, visitproc visit, void* arg)
+{
+#if PY_VERSION_HEX > 0x03090000
```
Comment (davisagli) [inline src/zope/interface/_zope_interface_coptimizations.c:1093]: check USE_HEAP_TYPES?

File: src/zope/interface/_zope_interface_coptimizations.c [hunk 4]
```diff
@@ -1784,296 +1779,873 @@ verifying_changed(verify *self, PyObject *ignored)
...
+/*
+ * Heap type: VerifyingBase
+ */
+static PyType_Slot VB_type_slots[] = {
+    {Py_tp_doc,         VB__doc__},
+    {Py_tp_traverse,    VB_traverse},
+    {Py_tp_clear,       VB_clear},
+    {Py_tp_dealloc,     VB_dealloc},
+    {Py_tp_methods,     VB_methods},
+    /* tp_base cannot be set as a stot -- pass to PyType_FromModuleAndSpec */
```
Comment (davisagli) [inline src/zope/interface/_zope_interface_coptimizations.c:1972]:
```suggestion
    /* tp_base cannot be set as a slot -- pass to PyType_FromModuleAndSpec */
```

File: src/zope/interface/_zope_interface_coptimizations.c [hunk 5]
```diff
@@ -1784,296 +1779,873 @@ verifying_changed(verify *self, PyObject *ignored)
...
-    return NULL;
+    /* our globals (exposed to Python) */
+    PyTypeObject*   specification_base_class;
+    PyTypeObject*   object_specification_descriptor_class;
+    PyTypeObject*   class_provides_base_class;
+    PyTypeObject*   interface_base_class;
+    PyTypeObject*   lookup_base_class;
+    PyTypeObject*   verifying_base_class;
+    PyObject*       adapter_hooks;
+    /* members importe from 'zope.interface.declarations'
```
Comment (davisagli) [inline src/zope/interface/_zope_interface_coptimizations.c:1999]:
```suggestion
    /* members imported from 'zope.interface.declarations'
```

File: src/zope/interface/_zope_interface_coptimizations.c [hunk 6]
```diff
@@ -1784,296 +1779,873 @@ verifying_changed(verify *self, PyObject *ignored)
...
+     */
+    PyObject*       empty;
+    PyObject*       fallback;
+    PyObject*       builtin_impl_specs;
+    PyTypeObject*   implements_class;
+    /* flag:  have we importe the next set of members yet from
```
Comment (davisagli) [inline src/zope/interface/_zope_interface_coptimizations.c:2005]:
```suggestion
    /* flag:  have we imported the next set of members yet from
```

File: src/zope/interface/_zope_interface_coptimizations.c (hunk 3 above)
Comment (tseaver) [reply src/zope/interface/_zope_interface_coptimizations.c:1101]: See above, and 25d664a

File: src/zope/interface/_zope_interface_coptimizations.c [hunk 7]
```diff
@@ -365,315 +263,447 @@ Spec_clear(Spec* self)
...
+    PyVarObject_HEAD_INIT(NULL, 0)
+    .tp_name           = SB__name__,
+    .tp_doc            = SB__doc__,
+    .tp_basicsize      = sizeof(SB),
+    .tp_flags          = BASETYPE_FLAGS,
+    .tp_call           = (ternaryfunc)SB__call__,
+    .tp_traverse       = (traverseproc)SB_traverse,
+    .tp_clear          = (inquiry)SB_clear,
+    .tp_dealloc        = (destructor)SB_dealloc,
+    .tp_weaklistoffset = offsetof(SB, weakreflist), /* XXX */
```
[the remark this answers is not in the document]
Comment (tseaver) [reply src/zope/interface/_zope_interface_coptimizations.c]: Actually, what I wanted was to remove the *manual* `weaklist` management, but the automated version is only available in Python >= 3.12.  6bb37a5 clarifies by defining a new `USE_EXPLICIT_WEAKREFLIST` flag, set true for Python < 3.12, and false for Python >= 3.12.  It also suppresses defining the `.tp_weaklistoffset` slot unless that flag is set.

File: src/zope/interface/_zope_interface_coptimizations.c [hunk 8]
```diff
@@ -334,15 +228,19 @@ typedef struct {
...
 /*
   We know what the fields are *supposed* to define, but
   they could have anything, so we need to traverse them.
 */
 static int
-Spec_traverse(Spec* self, visitproc visit, void* arg)
+SB_traverse(SB* self, visitproc visit, void* arg)
 {
+/* Visit our 'tp_type' only on Python >= 3.9 */
+#if PY_VERSION_HEX > 0x03090000
```
[the remark this answers is not in the document]
Comment (davisagli) [reply src/zope/interface/_zope_interface_coptimizations.c:258]: @tseaver I was figuring that the type only needs to be garbage-collected if it is a heap type, so we shouldn't visit it here if USE_HEAP_TYPES is 0, even when Python >= 3.9. That is currently the case for Python 3.9 and 3.10. But maybe my thinking here is wrong.

[the remark this answers is not in the document]
Comment (tseaver) [reply src/zope/interface/_zope_interface_coptimizations.c]: c35f95e

**Review by davisagli — APPROVED** [review]

**Comment by icemac** [comment]
These changes seem to break `zope.proxy`, see https://github.com/zopefoundation/z3c.pt/actions/runs/10110195498/job/28389675912?pr=30 – I can also create a new ticket for this issue if needed.

**Pull request merged**